Mathlib/Tactic/NormNum/Basic.lean

```diff
@@ -439,15 +439,10 @@ such that `norm_num` successfully recognises `a`. -/
 @[norm_num ¬_] def evalNot : NormNumExt where eval {u α} e := do
   let .app (.const ``Not _) (a : Q(Prop)) ← whnfR e | failure
   guard <|← withNewMCtxDepth <| isDefEq α q(Prop)
-  let .isBool b p ← derive q($a) | failure
-  haveI' : u =QL 0 := ⟨⟩; haveI' : $α =Q Prop := ⟨⟩
-  haveI' : $e =Q ¬ $a := ⟨⟩
```

Comment:
@gebner are these `haveI'` lines serving any purpose? You added them in #5975, and even at that time the code successfully typechecked without them.

Reply:
Ah, we are trying return a value of type `NormNum.Result e`, and we do it by analyzing `a`. The `$e =Q ¬ $a ` thing is so that the typechecker knows how `e` relates to `a`.

But then how does this code still work even when I delete these lines? Does that mean there's a bug in `Qq`?

Reply:
(I re-added the lines in https://github.com/leanprover-community/mathlib4/pull/9532/commits/27adfeb2d2e1907c586fb61842b6ff1873ec19d9, because this seems to me more likely to be a bug in Qq than a problem with the code here.)

Reply:
I would have expected that either `$e =Q ¬ $a` or the cast `have p : Q($a) := p` would be sufficient. But it seems from d94bd4270dc76ce714a56efb73ca4eb82178527a that we don't need either. So I'm in favour of removing these again.

On the other hand, I must admit that placating Qq is still a deeply magical process to me so I'm inclined to accept any variation of the code that compiles and looks plausible. So I'll let you decide.